File: 1424-diagonal-traverse-ii/1424-diagonal-traverse-ii.py
class Solution:
    def findDiagonalOrder(self, nums: List[List[int]]) -> List[int]:
        
        queue = collections.deque([(0,0)])
        ans = []
        
        while queue:
            size = len(queue)
            for k in range(size):
                i, j = queue.popleft()
                ans.append(abs(nums[i][j]))

                if i+1 < len(nums) and j < len(nums[i+1]) and nums[i+1][j] >= 0:
                    queue.append((i+1, j))
                    nums[i+1][j] = -1*nums[i+1][j]

                if j+1 < len(nums[i]) and nums[i][j+1] >= 0:
                    queue.append((i, j+1))
                    nums[i][j+1] = -1*nums[i][j+1]
        return ans

File: 1424-diagonal-traverse-ii/1424-diagonal-traverse-ii.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& nums) {
        queue<pair<int, int>> q;
        vector<int> res;

        q.push({0, 0});

        while(!q.empty()){
            int size = q.size();

            for(int i=0; i<size; i++){
                auto cur = q.front();
                q.pop();
                res.push_back(nums[cur.first][cur.second]);

                if(cur.first+1 < nums.size() && cur.second < nums[cur.first+1].size() && cur.second == 0){
                    q.push({cur.first+1, cur.second});
                }
                if(cur.second+1 < nums[cur.first].size()){
                    q.push({cur.first, cur.second+1});
                }
            }
        }
        return res;
    }
};